### PR TITLE
handle exception when VS2008 is not installed

### LINKED ISSUE #1048: conda-build does not handle absence of MSVC on windows well
I was trying to build the **math3d** package inside conda, and I am failing and getting the following error

C:\>conda skeleton pypi --recursive math3d --version 3.1.2 --output-dir C:\Anaconda2\skeleton
Using url https://files.pythonhosted.org/packages/1e/93/d0a462fc3acb5d9a71912129f9c72d448cd995ed98c632c729c6aa63d564/math3d-3.1.2.tar.gz (24 KB) for math3d.
Downloading math3d
Using cached download
Unpacking math3d...
done
working in c:\users\debang~1\appdata\local\temp\tmpsbez9sconda_skeleton_math3d-3.1.2.tar.gz
Fetching package metadata .........
Solving package specifications: ..........

The following NEW packages will be INSTALLED:

    mkl:    11.3.3-1
    numpy:  1.11.0-py27_1
    pyyaml: 3.11-py27_4
    yaml:   0.1.6-0

Linking packages ...
[      COMPLETE      ]|##################################################| 100%
Applying patch: 'c:\\users\\debang~1\\appdata\\local\\temp\\tmpsbez9sconda_skeleton_math3d-3.1.2.tar.gz\\pypi-distutils.patch'
patching file core.py
Hunk #1 succeeded at 169 with fuzz 2 (offset 2 lines).
Writing recipe for math3d
Done

C:\>conda build C:\Anaconda2\skeleton math3d
Removing old build environment
BUILD START: math3d-3.1.2-py27_0
Fetching package metadata .........
Solving package specifications: ..........

The following NEW packages will be INSTALLED:

    pip:            8.1.2-py27_0
    python:         2.7.11-4
    setuptools:     23.0.0-py27_0
    vs2008_runtime: 9.00.30729.1-1
    wheel:          0.29.0-py27_0

Linking packages ...
[      COMPLETE      ]|##################################################| 100%
Source cache directory is: C:\Anaconda2\conda-bld\src_cache
Found source in cache: math3d-3.1.2.tar.gz
Extracting download
Package: math3d-3.1.2-py27_0
source tree in: C:\Anaconda2\conda-bld\work\math3d-3.1.2
An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda-build/issues

Include the output of the command 'conda info' in your report.

Traceback (most recent call last):
  File "C:\Anaconda2\Scripts\conda-build-script.py", line 4, in <module>
    sys.exit(main())
  File "C:\Anaconda2\lib\site-packages\conda_build\main_build.py", line 144, in main
    args_func(args, p)
  File "C:\Anaconda2\lib\site-packages\conda_build\main_build.py", line 389, in args_func
    args.func(args, p)
  File "C:\Anaconda2\lib\site-packages\conda_build\main_build.py", line 332, in execute
    dirty=args.dirty, activate=args.activate)
  File "C:\Anaconda2\lib\site-packages\conda_build\build.py", line 563, in build
    import conda_build.windows as windows
  File "C:\Anaconda2\lib\site-packages\conda_build\windows.py", line 30, in <module>
    'installationfolder')
  File "C:\Anaconda2\lib\distutils\msvc9compiler.py", line 73, in get_value
    raise KeyError(key)
KeyError: 'installationfolder'

C:\>conda info
Current conda install:

             platform : win-64
        conda version : 4.1.3
    conda-env version : 2.5.1
  conda-build version : 1.21.2
       python version : 2.7.11.final.0
     requests version : 2.7.0
     root environment : C:\Anaconda2  (writable)
  default environment : C:\Anaconda2
     envs directories : C:\Anaconda2\envs
        package cache : C:\Anaconda2\pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/win-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/win-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : None
         offline mode : False
    is foreign system : False

### PR DESCRIPTION
**Pull request by msarahan**
Fixes #1048 - but needs testing.

### CODE REVIEW

**Comment by msarahan** [comment]
Tested this on a Win7 VM; found the extra exception (TypeError) was necessary.  Working ok now.  Merging.  Travis and Appveyor errors are from before, and are unaffected by this PR.

File: conda_build/windows.py [hunk 1]
```diff
@@ -154,6 +151,16 @@ def build_vcvarsall_cmd(cmd, arch=arch_selector):
         msvc_env_lines.append('if not "%WindowsSDKDir%" == "{}" ( {} )'.format(
             WIN_SDK_71_PATH, build_vcvarsall_cmd(vcvarsall_vs_path)))
     elif version == '9.0':
+        # Get the Visual Studio 2008 path (not the Visual C++ for Python path)
+        # and get the 'vcvars64.bat' from inside the bin (in the directory above
+        # that returned by distutils_find_vcvarsall)
+        try:
+            VCVARS64_VS9_BAT_PATH = os.path.join(os.path.dirname(distutils_find_vcvarsall(9)),
+                                                'bin', 'vcvars64.bat')
```
Comment (jakirkham) [inline conda_build/windows.py]: I'm confused. It looks like @patricksnape said this won't work in PR ( https://github.com/conda/conda-build/pull/1189 ) and had fixed it. Yet I'm seeing this showing up in [`master`]( https://github.com/conda/conda-build/blob/327d9a358e2b849a05e1cd33bb5f917dd96b9bb8/conda_build/windows.py#L169-L170 ) anyways. This is likely the cause of the issues [seen]( https://github.com/conda-forge/conda-smithy/pull/392#issuecomment-269534107 ) by @mwcraig . Was this intentional or just a bad merge?

**Comment by msarahan** [comment]
Wild guess: bad merge. Happy to take a look when I return to work next week.

**Comment by jakirkham** [comment]
Yeah, sorry, I may be confused as it seem @patricksnape left that line there in his PR too. 😕 So maybe it was intentional or overlooked. No idea.

Just trying to understand why Python 2.7 is not getting activated correctly.

**Comment by patricksnape** [comment]
Sorry, been totally offline for the last while! Anything I can do/clarify here?

**Comment by msarahan** [comment]
@patricksnape I'm not sure.  I think we haven't quite nailed down the case where VC compiler for Python 2.7 is installed, but VS 2008 Express (not pro) is also installed.  You and I handle that with the extra registry stuff, I think, but without that it fails.

As I understand it, the registry stuff should only be necessary for cmake.  Conda-build should be working fine without it.  Something is fishy that the conda-build test suite is working fine, but people are seeing errors elsewhere.

https://github.com/conda/conda-build/blob/master/tests/test_api_build.py#L270
https://github.com/conda/conda-build/blob/master/tests/test-recipes/metadata/_build_msvc_compiler/bld.bat#L10

**Comment by patricksnape** [comment]
Are people seeing errors on Appveyor or locally? It's really really hard to test when people have many variations of Visual Studio installed/not installed.

**Comment by msarahan** [comment]
IMHO, this goes back to your registry hack and bat file copying.  See https://github.com/conda/conda-build/issues/1606#issuecomment-270031876 for more info.  I'm not sure that there's a "good" conda-build fix here.  We can probably make it so that this fix isn't strictly necessary for things that don't use cmake, but that seems like covering up something that people will stumble on sooner or later.  The right answer is for people to install the ``vs2008_express_vc_python_patch`` package from conda-forge in their appveyor.yml - and probably for conda-forge to do that by default on feedstocks.

**Comment by patricksnape** [comment]
Hmm. What if cmake for Python 2.7 required the patch on Windows?

**Comment by msarahan** [comment]
It's a bad idea for anything to require the patch, because it requires
admin priveleges that we can't count on. =(

**Comment by patricksnape** [comment]
True - what if the preprocessor also detected the existence of the `CI` env variable, or the `APPVEYOR` one?

**Comment by msarahan** [comment]
That's a pretty good idea, along with some kind of fallback warning if that isn't found, and the compiler still fails.

**Comment by patricksnape** [comment]
Yes certainly a warning would go a long way with this one

**Pull request merged**